reader-privacy.md

```diff
@@ -0,0 +1,199 @@
+nor writers (hosts storing and distributing content) have much privacy with regard to content they publish or 
+consume. It is very easy for a content router node or a passive observer to learn which file is requested by 
+which client during the routing process, as the potential adversary easily learns about the requested `CID`. 
+A curious actor could request the same `CID` and download the associated file to monitor the user’s behavior. 
+This is obviously undesirable and has been for some time now a strong request from the community.
+
+The changes described in this specification introduce a IPNI Readres Privacy upgrade. It will prevent 
+passive observers from tracking user's actions as described above. It will also be a first step towards
+fully private IPNI protocol that will eliminate indexers as centralised observers.  
+
+### Non Goals
+
+* Writer, i.e. content provider or publisher, Privacy, which will be done in a separate specification
+* Retrieval Privacy, which is out of scope for the content routing subsystem.
+​
+## Background
+​
+Network indexers build their indexes by ingesting chains of Advertisements. Advertisement is a
+construct that allows Storage Providers to publish their CIDs in bulk (FIL deals) instead of doing
+that individually for each CID. A group of CIDs is represented by a unique ContextID as can be seen
+on the diagram below:
+​
+![Index building flow](resources/readers-privacy-1.png)
+
+## Specification
+​
+This specification improves the reader privacy by proposing changes to the Step 3, depicted above, where the client 
+supplies the content CID directly in order to lookup its corresponding providers.
+
+In order to protect the reader's privacy the proposal changes the way CID lookup works to the following:
+
+* A client who wants to do a lookup will calculate a hash over the CID's multihash (`hash(MH)`) and use it for the
+lookup query (hence the name double hashing);
+* In response to the hashed find request, the indexer will return a set of encrypted `ProviderRecordKey`s. 
+`ProviderRecordKey` will consist of the `peerID` concatenated with a hash over `contextID`. 
+Each `ProviderRecordKey` will be encrypted with a key derived from the *original* multihash value: 
+`enc(peerID || hash(contextID), MH)`, where `hash` is a hash over the value, and `||` is concatenation 
+and `enc` is encryption over the value. In order to make sense of that payload, a passive observer would need 
+to get hold of the original CID that isn't revealed during the communication round;
+* Using the original multihash, the client will decrypt `ProviderRecordKey`s and then calculate a hash
+over the decrypted `peerID` part of it. Using such hash for each `ProviderRecordKey` the client would do another lookup 
+to get an encrypted `ProviderRecord` in response. `ProviderRecord` will contain information about the provider, 
+such as it's *peerID*, *multiaddrs*, *supported protocols* and so on. Each `ProviderRecord` will be encrypted 
```

Comment:
How does the lookup for `ProviderRecord` work if the lookup key consists of the hash of `peerID` only? `ProviderRecord` is uniquely identified by `PeerID` _and_ `ContextID`; right?

Is the idea that the server will send back the list of _all_ `ProviderRecord`s for the given key and the client would filter by `ContextID` on their side to avoid revealing the `ContextID` to the server?

Reply:
Great question. By `ProviderRecord` here I mean a tuple of `peerID` to a list of `multiaddresses` (i.e. without the deal-specific metadata). I will additionally clarify this in the spec.

Reply:
Thank you; what about context ID? 

It looks like the returned `hash(contextID)` is not used at all. How would the client retrieve the IPNI `Metadata`?

Reply:
The idea is to leave it for future improvements. In next iteration we can make `hash(contextID)` to be an optional input parameter that would add context-specific data to the returned payload.